### PR TITLE
Fix for the latest libraries

### PR DESCRIPTION
**Pull request by ryosu-sato**
I have managed to build r_type.
(but, have not checked the regression)

### CODE REVIEW

File: src/data/cond.ml [hunk 1]
```diff
@@ -64,9 +64,9 @@ module Operation = struct
   let false_ = Value Objt.false_
   let and_ x y = op2 x Op.And_ y
   let and_s x y =
-    if x = Value (Objt.true_) then
+    if Stdlib.(x = Value) (Objt.true_) then
```
Comment (AdrienChampion) [inline src/data/cond.ml]: This does not compile on my machine, I fixed it with

```diff
- if Stdlib.(x = Value) (Objt.true_) then
+ if Stdlib.(x = Value Objt.true_) then
```

My fix seems fine to me, but you said you compiled the project and I don't understand how given this line.

File: src/data/cond.ml [hunk 2]
```diff
@@ -64,9 +64,9 @@ module Operation = struct
...
       y
-    else if y = Value (Objt.true_) then
+    else if Stdlib.(y = Value) (Objt.true_) then
```
Comment (AdrienChampion) [inline src/data/cond.ml]: Same

```diff
- else if Stdlib.(y = Value) (Objt.true_) then
+ else if Stdlib.(y = Value Objt.true_) then
```

**Review by AdrienChampion — COMMENTED** [review]
@ryosu-sato I see two problems I managed to fix, I think, see below.

But it still does not compile on my machine:

```
❯ omake
*** omake: reading OMakefiles
*** omake: finished reading OMakefiles (0.01 sec)
- build src/loader fmlparser.ml
+ menhir --ocamlc 'ocamlfind ocamlc -package core,ppx_deriving,ppx_deriving.create,ppx_deriving.eq,ppx_deriving.ord,ppx_deriving.show,ppx_sexp_conv,ppx_hash,ppx_variants_conv,ppx_fields_conv,ppx_compare -warn-error A -thread -g -I ../common -I ../concern -I ../data -I ../mlLoader' fmlparser.mly
File "fmlparser.mly", line 15, characters 95-101:
Warning: the token ASSERT is unused.
File "fmlparser.mly", line 15, characters 60-64:
Warning: the token ELSE is unused.
File "fmlparser.mly", line 15, characters 90-94:
Warning: the token FAIL is unused.
File "fmlparser.mly", line 15, characters 74-78:
Warning: the token FUNC is unused.
File "fmlparser.mly", line 15, characters 51-53:
Warning: the token IF is unused.
File "fmlparser.mly", line 15, characters 87-89:
Warning: the token IN is unused.
File "fmlparser.mly", line 15, characters 66-69:
Warning: the token LET is unused.
File "fmlparser.mly", line 15, characters 70-73:
Warning: the token REC is unused.
File "fmlparser.mly", line 15, characters 102-108:
Warning: the token RECAND is unused.
File "fmlparser.mly", line 15, characters 109-115:
Warning: the token SELECT is unused.
File "fmlparser.mly", line 15, characters 54-58:
Warning: the token THEN is unused.
File "fmlparser.mly", line 27, characters 10-16:
Warning: the token UMINUS is unused.
File "fmlparser.mly", line 15, characters 116-120:
Warning: the token WHEN is unused.
File "fmlparser.mly", line 17, characters 0-5:
Warning: the precedence level assigned to ASSERT is never useful.
File "fmlparser.mly", line 27, characters 0-9:
Warning: the precedence level assigned to UMINUS is never useful.
Warning: 19 states have shift/reduce conflicts.
Warning: one state has reduce/reduce conflicts.
Warning: 33 shift/reduce conflicts were arbitrarily resolved.
- build src/horn_solver horn_solver.cmxa
+ ocamlfind ocamlopt -package core -warn-error A -thread -inline 100 -inline 100 -a -o horn_solver.cmxa parseBase.cmx SmtParse.cmx smtLex.cmx solver.cmx
File "_none_", line 1:
Error: Cannot find file SmtParse.cmx
*** omake: 222/324 targets are up to date
*** omake: failed (6.39 sec, 5/48 scans, 29/95 rules, 92/150 digests)
*** omake: targets were not rebuilt because of errors:
   src/horn_solver/horn_solver.a
   src/horn_solver/horn_solver.cmxa
```

# Opam/OCaml/Libs Version Info

<details>

```
❯ ocamlc --version
4.11.1

❯ opam --version
2.0.5

❯ opam list -i
# Packages matching: installed
# Name                  # Installed  # Synopsis
base                    v0.14.1      Full standard library replacement for OCaml
base-bigarray           base
base-threads            base
base-unix               base
base_bigstring          v0.14.0      String type based on [Bigarray], for use in I/O and C-bindings
base_quickcheck         v0.14.0      Randomized testing framework, designed for compatibility with Base
bin_prot                v0.14.0      A binary protocol generator
conf-g++                1.0          Virtual package relying on the g++ compiler (for C++)
conf-m4                 1            Virtual package relying on m4
core                    v0.14.0      Industrial strength alternative to OCaml's standard library
core_kernel             v0.14.1      Industrial strength alternative to OCaml's standard library
cppo                    1.6.7        Code preprocessor like cpp for OCaml
csexp                   1.3.2        Parsing and printing of S-expressions in Canonical form
dune                    2.8.1        Fast, portable, and opinionated build system
dune-configurator       2.8.1        Helper library for gathering system configuration
fieldslib               v0.14.0      Syntax extension to define first class values representing record fields, to get and set record fields, iterate and fold over all fields of a record and create new record values
jane-street-headers     v0.14.0      Jane Street C header files
jbuilder                1.0+beta20.2 Fast, portable and opinionated build system
jst-config              v0.14.0      Compile-time configuration for Jane Street libraries
menhir                  20201216     An LR(1) parser generator
menhirLib               20201216     Runtime support library for parsers generated by Menhir
menhirSdk               20201216     Compile-time library for auxiliary tools related to Menhir
num                     1.4          The legacy Num library for arbitrary-precision integer and rational arithmetic
ocaml                   4.11.1       The OCaml compiler (virtual package)
ocaml-base-compiler     4.11.1       Official release 4.11.1
ocaml-compiler-libs     v0.12.3      OCaml compiler libraries repackaged
ocaml-config            1            OCaml Switch Configuration
ocaml-migrate-parsetree 1.8.0        Convert OCaml parsetrees between different versions
ocamlfind               1.8.1        A library manager for OCaml
octavius                1.2.2        Ocamldoc comment syntax parser
omake                   0.10.3       Build system designed for scalability and portability
parsexp                 v0.14.0      S-expression parsing library
ppx_assert              v0.14.0      Assert-like extension nodes that raise useful errors on failure
ppx_base                v0.14.0      Base set of ppx rewriters
ppx_bench               v0.14.1      Syntax extension for writing in-line benchmarks in ocaml code
ppx_bin_prot            v0.14.0      Generation of bin_prot readers and writers from types
ppx_cold                v0.14.0      Expands [@cold] into [@inline never][@specialise never][@local never]
ppx_compare             v0.14.0      Generation of comparison functions from types
ppx_custom_printf       v0.14.0      Printf-style format-strings for user-defined string conversion
ppx_derivers            1.2.1        Shared [@@deriving] plugin registry
ppx_deriving            5.1          Type-driven code generation for OCaml
ppx_enumerate           v0.14.0      Generate a list containing all values of a finite type
ppx_expect              v0.14.0      Cram like framework for OCaml
ppx_fields_conv         v0.14.1      Generation of accessor and iteration functions for ocaml records
ppx_fixed_literal       v0.14.0      Simpler notation for fixed point literals
ppx_hash                v0.14.0      A ppx rewriter that generates hash functions from type expressions and definitions
ppx_here                v0.14.0      Expands [%here] into its location
ppx_inline_test         v0.14.1      Syntax extension for writing in-line tests in ocaml code
ppx_jane                v0.14.0      Standard Jane Street ppx rewriters
ppx_js_style            v0.14.0      Code style checker for Jane Street Packages
ppx_let                 v0.14.0      Monadic let-bindings
ppx_module_timer        v0.14.0      Ppx rewriter that records top-level module startup times
ppx_optcomp             v0.14.0      Optional compilation for OCaml
ppx_optional            v0.14.0      Pattern matching on flat options
ppx_pipebang            v0.14.0      A ppx rewriter that inlines reverse application operators `|>` and `|!`
ppx_sexp_conv           v0.14.1      [@@deriving] plugin to generate S-expression conversion functions
ppx_sexp_message        v0.14.0      A ppx rewriter for easy construction of s-expressions
ppx_sexp_value          v0.14.0      A ppx rewriter that simplifies building s-expressions from ocaml values
ppx_stable              v0.14.1      Stable types conversions generator
ppx_string              v0.14.1      Ppx extension for string interpolation
ppx_typerep_conv        v0.14.1      Generation of runtime types from type declarations
ppx_variants_conv       v0.14.1      Generation of accessor and iteration functions for ocaml variant types
ppxlib                  0.15.0       Standard library for ppx rewriters
re                      1.9.0        RE is a regular expression library for OCaml
re2                     v0.14.0      OCaml bindings for RE2, Google's regular expression library
result                  1.5          Compatibility Result module
seq                     base         Compatibility package for OCaml's standard iterator type starting from 4.07.
sexplib                 v0.14.0      Library for serializing OCaml values to and from S-expressions
sexplib0                v0.14.0      Library containing the definition of S-expressions and some base converters
spawn                   v0.13.0      Spawning sub-processes
splittable_random       v0.14.0      PRNG that can be split into independent streams
stdio                   v0.14.0      Standard IO library for OCaml
stdlib-shims            0.1.0        Backport some of the new stdlib features to older compiler
time_now                v0.14.0      Reports the current time
timezone                v0.14.0      Time-zone handling
typerep                 v0.14.0      Typerep is a library for runtime types
variantslib             v0.14.0      Part of Jane Street's Core library
```
</details>

File: src/data/cond.ml (hunk 2 above)
Comment (ryosu-sato) [reply src/data/cond.ml]: Oops, I pushed the wrong commit.
Your fix seems to be correct, and I have pushed it to my repository.

I think the failure is caused by some dependency problem.
Could you try to recompile it after the command "omake clean", or in a re-cloned directory?

**Comment by AdrienChampion** [comment]
@ryosu-sato Still does not compile on my machine. Can you merge with `master` (or `build_update`)? This should trigger CI on your PR so that we can see the problems directly.

If `r_type` does compile on your machine, can you edit `.github/workflows/setup.sh` so that CI uses the right library versions?

It would be great if we could compile on the latest compiler. If that's not the one you are using, please make sure `.github/workflows/setup.sh` or `ci.yml` use the same compiler you are.

**Comment by ryosu-sato** [comment]
Here is the result of CI.
[https://github.com/ryosu-sato/r_type/actions/runs/497493000](https://github.com/ryosu-sato/r_type/actions/runs/497493000).

The build succeded in Ubuntu, but failed in macOS.
I cannot inspect the failure, since I do not have a macOS device now.

**Comment by ryosu-sato** [comment]
I changed r_type to use Dune build system instead of OMake.
The change enables a build on macOS.
[https://github.com/ryosu-sato/r_type/actions/runs/559692776](https://github.com/ryosu-sato/r_type/actions/runs/559692776)

**Comment by AdrienChampion** [comment]
Works on my machine too, thanks!

**Pull request merged**